src/vision/shiny_future/grace_writes_a_recursive_fn_with_auto_box.md

```diff
@@ -0,0 +1,119 @@
+# ✨ Shiny future stories: Grace writes a recursive function with auto-box
+
+## 🚧 Warning: Draft status 🚧
+
+This is a draft "shiny future" story submitted as part of the brainstorming period. It is derived from what actual Rust users wish async Rust should be, and is meant to deal with some of the challenges that Async Rust programmers face today.
+
+If you would like to expand on this story, or adjust the answers to the FAQ, feel free to open a PR making edits (but keep in mind that, as peoples needs and desires for async Rust may differ greatly, shiny future stories [cannot be wrong]. At worst they are only useful for a small set of people or their problems might be better solved with alternative solutions). Alternatively, you may wish to [add your own shiny vision story][htvsq]!
+
+## The story
+
+### Early days
+
+Grace is using async Rust for the first time. She is experimenting with async functions and she begins with her favorite "Hello, World" sort of application, Fibonacci:
+
+```rust
+async fn fibonacci(n: u32) -> u32 {
+    match n {
+        0 => 1,
+        1 => 1,
+        _ => fibonacci(n - 1) + fibonacci(n - 2),
+    }
+}
+```
+
+It doesn't make much sense, but that's ok. She compiles it as part of a webapp, so that when she requests `https://localhost/fib/22` she gets back a page with the correct result. Fun!
+
+### Discovering automatic boxing
+
+Now that Grace has the basics working, she starts adding code to the [MonsterMesh] project that she is developing at work. She is writing a simple recursive function, similar to `fibonacci`, but she finds that she gets an error:
+
+```
+error: `Box` required for async function
+ --> src/lib.rs:4:9
+  |
+4 |     async fn fibonacci(x: usize) -> usize {
+  |              ^^^^^^^^^ the function `fibonacci` is recursive and requires a `Box`
+  |
+note: the lint level is defined here
+ --> src/lib.rs:1:9
+  |
+1 | #![deny(automatic_box)]
+  |         ^^^^^^^^^^^^^
+help: to request a box explicitly, write `box async fn`
+  |
+4 |     box async fn fibonacci(x: usize) -> usize {
+```
+
+"What's this?" she thinks. "It worked before?" She clicks in her IDE on the "more info" page for the `automatic_box` lint, where she learns that recursive asynchronous functions require some form of extra heap allocation to work. The compiler normally inserts these automatically, but there is an allow-by-default lint that can be used to turn those into warnings or hard errors.
```

Comment:
So, if the lint hadn't been there, the heap allocation would have been added even if Grace hadn't added the explicit `box` annotation? It seems like we'd more likely want to error automatically but add an error message explaining exactly what to do to get the code compiling.

Reply:
I disagree, but feel free to write your own story :)

Reply:
I wanted to write a version where people don't have to care about memory allocations, but they have the option to do so. I would however be happy to add some parts of the story describing some of the various downsides one might see from this scenario!

Reply:
(I would though like to read the story where they *do* have to care, as well.)